Add Z value support to geo_shape

nknize opened the pull request:
This enhancement adds Z value support (source only) to geo_shape fields. If vertices are provided with a third dimension, the third dimension is ignored for indexing but returned as part of source. Like before, any values greater than the 3rd dimension are ignored.

closes #22917

jpountz commented on core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java:
```diff
@@ -170,16 +175,24 @@ public static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper geoD
     protected static void writeCoordinateTo(Coordinate coordinate, StreamOutput out) throws IOException {
         out.writeDouble(coordinate.x);
         out.writeDouble(coordinate.y);
+        out.writeOptionalDouble(Double.isNaN(coordinate.z) ? null : coordinate.z);
```
you need to add a version check here?

jpountz commented on core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java:
```diff
@@ -170,16 +175,24 @@ public static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper geoD
-        return new Coordinate(in.readDouble(), in.readDouble());
+        double x = in.readDouble();
+        double y = in.readDouble();
+        Double z = in.readOptionalDouble();
```
you need to add a version check here?

jpountz commented on core/src/test/java/org/elasticsearch/common/geo/ShapeBuilderTests.java:
```diff
@@ -650,4 +650,49 @@ public void testInvalidShapeWithConsecutiveDuplicatePoints() {
+            .coordinate(new Coordinate(-45, -30, 101))
+            .coordinate(new Coordinate(-45, 30, 110)));
+
+        assertEquals(expected, pb.toString());
```
we should test parsing too?

nknize replied:
added `testParse3DPolygon` to `GeoJSONShapeParserTests`

hanoch commented:
@nknize #22917 talks about both geo_point and geo_shape. does this PR include support for store-and-retrieve Z values for geo_point as well, or will that support be separate from this PR?

nknize commented:
@hanoch `geo_point` already supports Z values in source. 

![geopoint_stored_z](https://user-images.githubusercontent.com/830187/28892923-35c025e4-7795-11e7-8523-2d818492a713.png)


This PR effectively fixes a bug where Z values for `geo_shape` coordinates were omitted in source after indexing.

javanna commented:
@nknize are you still working on this?

hanoch commented:
@nknize we are still waiting for this functionality to be merged into the master branch.

hanoch commented:
cc: @sherry-ger

nknize commented:
@colings86 This should be good to go. Could you give it a final once over since @jpountz is out? Thx

clintongormley commented:
Discussed this in FixItFriday.  We're not crazy about accepting parameters that we will silently ignore, but given that Z values are part of the geojson spec, we should probably make an effort to extract what info we can from valid geojson files.

It should, however, be enabled by opting in using a mapping parameter like `accept_but_ignore_z_values`, which makes clear to the user what they are opting in to.

hanoch commented:
are we finally ready to merge this to master anytime soon?

hanoch commented:
@colings86 @nknize is this now ready to be merged to master to be able to address #22917?

nknize commented:
Rebased to master and cleaned up some geo_point parsing. This is ready for another review and should be just about ready to merge. /cc @jpountz @colings86

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/GeoPointFieldMapper.java:
```diff
@@ -57,11 +60,13 @@
     public static class Defaults {
         public static final Explicit<Boolean> IGNORE_MALFORMED = new Explicit<>(false, false);
         public static final GeoPointFieldType FIELD_TYPE = new GeoPointFieldType();
+        public static final Explicit<Boolean> ACCEPT_Z_VALUE = new Explicit<>(true, false);
```
true by default? shouldn't it be false like the docs say? (we might need some logic to make it true for pre-6.3 indices in case we used to accept it in the past)

nknize commented on server/src/main/java/org/elasticsearch/common/geo/builders/CoordinatesBuilder.java:
```diff
@@ -41,7 +42,12 @@
-        this.points.add(coordinate);
+        if (points.isEmpty() == false && Double.isNaN(points.get(0).z) != Double.isNaN(coordinate.z)) {
+            throw new ElasticsearchException("unable to add coordinate to CoordinateBuilder: " +
+                "coordinate dimensions do not match");
```
:+1:

nknize commented on docs/reference/mapping/types/geo-point.asciidoc:
```diff
@@ -105,6 +105,13 @@ The following parameters are accepted by `geo_point` fields:
     If `true`, malformed geo-points are ignored. If `false` (default),
     malformed geo-points throw an exception and reject the whole document.
 
+<<accept_z_value,`accept_z_value`>>::
```
:+1:

nknize commented on server/src/main/java/org/elasticsearch/common/Strings.java:
```diff
@@ -396,6 +396,40 @@ public static boolean validFileNameExcludingAstrix(String fileName) {
+     * @param ch  the char to count
+     * @return the number of occurrences, 0 if the CharSequence is {@code null}
+     */
+    public static int countMatches(final CharSequence str, final char ch) {
```
:+1:

nknize commented on server/src/main/java/org/elasticsearch/common/geo/GeoPoint.java:
```diff
@@ -75,14 +79,23 @@ public GeoPoint resetLon(double lon) {
+            String[] vals = value.split(",");
+            double lat = Double.parseDouble(vals[0].trim());
+            double lon = Double.parseDouble(vals[1].trim());
+            double alt = Double.NaN;
```
:+1:

nknize commented on server/src/main/java/org/elasticsearch/common/geo/builders/CircleBuilder.java:
```diff
@@ -173,6 +173,10 @@ public String toWKT() {
     }
 
+    public int numDimensions() {
+        return center.z == Double.NaN ? 2 : 3;
```
ah, yes... thank you. I still slip up w/ this one occasionally.

jpountz commented on server/src/main/java/org/elasticsearch/common/geo/GeoPoint.java:
```diff
@@ -75,14 +77,20 @@ public GeoPoint resetLon(double lon) {
+
+    public GeoPoint resetFromString(String value, final boolean ignoreZValue) {
+        if (value.contains(",")) {
+            String[] vals = value.split(",");
```
please fail if vals.length > 3

hanoch replied:
Wouldn't it better to be more lenient and accept the first 3 (or 2) dimensions, rather than completely fail? Adding failure now will break compatibility with previous behavior.

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java:
```diff
@@ -100,6 +100,7 @@
         public static final double LEGACY_DISTANCE_ERROR_PCT = 0.025d;
         public static final Explicit<Boolean> COERCE = new Explicit<>(false, false);
         public static final Explicit<Boolean> IGNORE_MALFORMED = new Explicit<>(false, false);
+        public static final Explicit<Boolean> ACCEPT_Z_VALUE = new Explicit<>(true, false);
```
s/ACCEPT/IGNORE/

jpountz commented on server/src/main/java/org/elasticsearch/common/geo/parsers/GeoJsonParser.java:
```diff
@@ -136,36 +143,44 @@ protected static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper s
+        }
         while (token == XContentParser.Token.VALUE_NUMBER) {
             token = parser.nextToken();
         }
```
let's remove this leniency?

hanoch commented:
@jpountz @nknize what is the overall design goal wrt the default behavior with Z values with geo_shape and geo_point?
It would be best if the default setting and behavior will continue to be that if the data value (csv-string/array/GeoJson/etc.) has a third Z dimension - it will be stored without needing to set any special settings. Also, it will be best to continue to auto ignore any 4th (e.g. M value) and above dimension in the data by default (without adjusting setting to accept the data) and store the first 3 (or 2 if the Z value is not present) dimensions.
That behavior will be consistent with the existing geo_point [x,y,z] or [x,y] array data values (and we should match the csv-string values with a Z value "y,x,,z" or "y,x" as well) of existing and previous releases, where if the data has a Z value it will be stored, without needing to set some special settings, or modify the default settings.
Is that the design?

colings86 commented:
@hanoch we do not want the Z Values to be accepted by default because it can be surprising to users. The fact that the z value is accepted but is not indexed and therefore does not accept search results is not what a user would expect OOTB so we feel it is a better user experience for the user to opt-in to accepting but ignoring the Z-Value.

hanoch commented:
is there anything else pending or are we ready to merge this PR?
